rust/feature-flags/src/flag_matching.rs

```diff
@@ -32,73 +166,292 @@ pub struct Person {
+    pub database_client: Option<DatabaseClientArc>,
     person_property_overrides: Option<HashMap<String, Value>>,
-    // TODO handle group properties
-    // group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+    group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+    group_type_mapping_cache: Arc<GroupTypeMappingCache>,
+    properties_cache: Arc<RwLock<PropertiesCache>>,
+    #[cfg(test)]
+    test_hash: Option<f64>,
 }
 
 const LONG_SCALE: u64 = 0xfffffffffffffff;
 
 impl FeatureFlagMatcher {
     pub fn new(
         distinct_id: String,
-        database_client: Option<Arc<dyn DatabaseClient + Send + Sync>>,
+        team_id: TeamId,
+        database_client: Option<DatabaseClientArc>,
         person_property_overrides: Option<HashMap<String, Value>>,
-        // group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+        group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+        group_type_mapping_cache: Option<Arc<GroupTypeMappingCache>>,
+        properties_cache: Option<Arc<RwLock<PropertiesCache>>>,
     ) -> Self {
         FeatureFlagMatcher {
-            // flags,
             distinct_id,
-            database_client,
-            cached_properties: None,
+            team_id,
+            database_client: database_client.clone(),
             person_property_overrides,
-            // group_property_overrides,
+            group_property_overrides,
+            group_type_mapping_cache: group_type_mapping_cache
+                .unwrap_or_else(|| Arc::new(GroupTypeMappingCache::new(team_id, database_client))),
+            properties_cache: properties_cache
+                .unwrap_or_else(|| Arc::new(RwLock::new(PropertiesCache::default()))),
+            #[cfg(test)]
+            test_hash: None,
         }
     }
 
-    pub async fn get_match(
+    /// Evaluate feature flags for a given distinct_id
+    /// - Returns a map of feature flag keys to their values
+    /// - If an error occurs while evaluating a flag, it will be logged and the flag will be omitted from the result
+    pub async fn evaluate_feature_flags(
         &mut self,
-        feature_flag: &FeatureFlag,
-    ) -> Result<FeatureFlagMatch, FlagError> {
-        if self.hashed_identifier(feature_flag).is_none() {
+        feature_flags: FeatureFlagList,
+    ) -> FlagsResponse {
+        let mut result = HashMap::new();
+        let mut error_while_computing_flags = false;
+
+        // Step 2: Fetch and cache all relevant person and group properties
+        // If we fail to fetch properties, we'll return an error response immediately
+        if let Err(e) = self
+            .fetch_and_cache_properties(
+                &feature_flags
+                    .flags
+                    .iter()
+                    .filter_map(|flag| flag.get_group_type_index())
+                    .collect(),
+            )
+            .await
+        {
```

Comment:
hmm, didn't read the entire code, but this is not ideal, because not all flags depend on properties (most experiment flags don't), and when the db is down for whatever reason, we don't want to error out immediately - we want to do a best effort match first for flags that can be computed.

Reply:
(hence the point of first pass doing a partial match with only override properties, and then fetching and doing a full match)

Reply:
oh got it, I didn't realize that experiment flags would manifest as override properties.  Your comment makes sense though!  I'll change my approach.